server/src/main/java/org/opensearch/tracing/DefaultTracer.java

```diff
@@ -0,0 +1,229 @@
+import io.opentelemetry.context.Context;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ *
+ * The default tracer implementation. This class implements the basic logic for span lifecycle and its state management.
+ * It also handles tracing context propagation between spans.
+ *
+ * It internally uses OpenTelemetry tracer.
+ *
+ */
+public class DefaultTracer implements Tracer {
+
+    public static final String CURRENT_SPAN = "current_span";
+
+    private static final Logger logger = LogManager.getLogger(DefaultTracer.class);
+    private static final String TRACE_ID = "trace_id";
+    private static final String SPAN_ID = "span_id";
+    private static final String SPAN_NAME = "span_name";
+    private static final String PARENT_SPAN_ID = "p_span_id";
+    private static final String THREAD_NAME = "th_name";
+    private static final String PARENT_SPAN_NAME = "p_span_name";
+    private static final String ROOT_SPAN = "RootSpan";
+
+    private final ThreadPool threadPool;
+    private final TracerSettings tracerSettings;
+    private final io.opentelemetry.api.trace.Tracer otelTracer;
+    private final OpenTelemetry openTelemetry;
+
+    public DefaultTracer(OpenTelemetry openTelemetry, ThreadPool threadPool, TracerSettings tracerSettings) {
+        this.openTelemetry = openTelemetry;
+        this.otelTracer = openTelemetry.getTracer("os-tracer");
+        this.threadPool = threadPool;
+        this.tracerSettings = tracerSettings;
+    }
+
+    @Override
+    public void startSpan(String spanName, Level level) {
+        Span span = createSpan(spanName, getCurrentSpan(), level);
+        setCurrentSpanInContext(span);
+        setSpanAttributes(span);
+    }
+
+    @Override
+    public void endSpan() {
+        Span currentSpan = getCurrentSpan();
+        if (currentSpan != null) {
+            endSpan(currentSpan);
+            setCurrentSpanInContext(currentSpan.getParentSpan());
+        }
+    }
+
+    @Override
+    public void addAttribute(String key, String value) {
+        addSingleAttribute(AttributeKey.stringKey(key), value);
+    }
+
+    @Override
+    public void addAttribute(String key, long value) {
+        addSingleAttribute(AttributeKey.longKey(key), value);
+
+    }
+
+    @Override
+    public void addAttribute(String key, double value) {
+        addSingleAttribute(AttributeKey.doubleKey(key), value);
+    }
+
+    @Override
+    public void addAttribute(String key, boolean value) {
+        addSingleAttribute(AttributeKey.booleanKey(key), value);
+    }
+
+    @Override
+    public void addEvent(String event) {
+        Span currentSpan = getCurrentSpan();
+        if (currentSpan instanceof OSSpan && ((OSSpan) currentSpan).getOtelSpan() != null) {
```

Comment:
for NoOpScan, shouldnt this early exit from here?

Reply:
NoopSpan would fail instanceOf check, so it should be ok I believe